### PR TITLE
[CUDA] Add high-performance Top-K kernels and online benchmarking

### PR DESCRIPTION
**Pull request by tianleiwu**
This introduces a major enhancement to the CUDA Top-K implementation, directly targeting the performance-critical task of token sampling in Large Language Models. It replaces a one-size-fits-all approach with a sophisticated framework that includes multiple specialized kernels and an intelligent, self-tuning dispatcher.

The existing full_sort and selection_sort kernels are not optimal across the wide range of problem sizes (batch size, vocab size, k) encountered in generation. This PR adds four new high-performance kernels and an online benchmarking system to dynamically select the best one at runtime.

### 1. Specialized Top-K Kernels
Four new Top-K algorithms have been added, each with its own strengths:

**per_batch_radix_sort**: Implements a per-batch sort using CUB's DeviceRadixSort. It is highly efficient for small batch sizes where launching independent sorts incurs low overhead.

**hybrid_sort**: A portable, multi-kernel, host-planned hybrid Top-K algorithm. This algorithm is designed for portability and does not require cooperative launch. It uses a host-side planner to orchestrate a multi-stage reduction process. It could support large batch size and k up to 256. This kernel could support vocab_size up to 1,245,184.

**iterative_reduce_sort**: A high-performance, single-kernel implementation that uses CUDA Cooperative Groups for grid-wide synchronization. It performs an iterative tree-based reduction to find the global top K, offering excellent performance when supported by the hardware.

**cascaded_reduce_sort**: A specialized version of iterative_reduce_sort, purpose-built for common LLM vocabulary sizes. Instead of an iterative reduction, it uses a fixed, up to three-step reduction cascade. By targeting typical LLM problem shapes, it often outperforms the above algorithms. This kernel could support vocab_size up to 311,296.

**flash_convergent_sort**: A flash convergent sort uses a one-step reduction. It is simple and fast. It often outperforms all other algorithms on k from range of 16 to 64. This kernel could support vocab_size up to 311,296.

### 2. Online Benchmarking and Dynamic Dispatch
The core of this PR is the new RunTopK dispatcher. Instead of being hard-coded to a single algorithm, it intelligently selects the fastest kernel for any given problem configuration.

How it works:

First Run: On the first call with a specific (batch_size, vocab_size, k) combination, the dispatcher runs a quick, lightweight, in-line benchmark of all applicable algorithms.

Caching: The winning algorithm is cached globally. This ensures that the benchmarking cost is paid only once per problem shape. Subsequent calls have zero overhead.

Dynamic Dispatch: All future calls with the same parameters will bypass the benchmark and directly invoke the fastest known kernel, ensuring optimal performance automatically.

This "just-in-time" optimization strategy guarantees that the best possible kernel is always used, adapting to different hardware, driver versions, and workloads without any need for manual tuning.

## Performance Benchmark Results (NVIDIA RTX 4090)
The new kernels provide a substantial performance improvement. The cascaded_reduce_sort and flash_convergent_sort kernels are consistently the fastest for typical LLM inference scenarios (small batch, large vocabulary).

Here is a summary of benchmark results comparing the new algorithms. The benchmark results for vocab_size=201088 are from UBuntu 24.04 in WSL with CUDA 12.8 and Nvidia RTX 4090. Latencies are reported in microseconds (µs), and speedup is relative to baseline (distributed select sort for batch size=1 and k <= 64, select sort for batch size >1 and k <= 64, full sort in other case).

The alias of algorithm is used in the below tables:
* full : segmented radix sort
* radix: per batch radix sort
* d-select: distributed select sort
* flash: flash convergent sort
* hybrid: multi kerel reduce sort
* iterative: iterative reduce sort
* cascaded: cascaded reduce sort

### batch\_size=1, vocab\_size=201088

| k | full | radix | select | d-select | flash | hybrid | iterative | cascaded | best | Speed Up |
|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 1103.9 | 108.5 | 13.7 | 9.4 | 17.5 | 18.7 | 17.9 | 17.4 | Distributed | 1.0x |
| 2 | 1065.0 | 105.5 | 27.1 | 12.8 | 17.2 | 18.7 | 18.1 | 17.6 | Distributed | 1.0x |
| 4 | 1065.2 | 111.1 | 48.3 | 16.6 | 17.1 | 18.9 | 18.1 | 17.3 | Distributed | 1.0x |
| 5 | 1065.6 | 105.1 | 58.5 | 19.0 | 18.0 | 19.6 | 18.1 | 18.2 | Flash | 1.1x |
| 8 | 1065.9 | 128.9 | 89.8 | 26.4 | 17.9 | 19.9 | 18.3 | 18.3 | Flash| 1.5x |
| 16 | 1064.6 | 109.2 | 173.6 | 45.6 | 18.7 | 24.8 | 20.3 | 19.4 | Flash | 2.4x |
| 32 | 1068.4 | 103.2 | 342.1 | 85.6 | 20.6 | 23.6 | 22.2 | 22.0 | Flash | 4.2x |
| 50 | 1065.9 | 102.8 | 530.6 | 131.3 | 22.6 | 25.8 | 25.2 | 24.9 | Flash | 5.8x |
| 64 | 1065.2 | 104.9 | 677.9 | 167.5 | 23.7 | 24.5 | 25.0 | 24.8 | Flash | 7.1x |
| 100 | 1065.5 | 107.4 | 1056.5 | NA | NA | 27.7 | NA | NA | Hybrid| 38.5x |
| 128 | 1065.9 | 109.3 | 1351.5 | NA | NA | 28.1 | NA | NA | Hybrid| 37.9x |
| 256 | 1063.4 | 112.1 | 2691.0 | NA | NA | 31.4 | NA | NA | Hybrid| 33.8x |

### batch\_size=4, vocab\_size=201088

| k | full | radix | select | d-select| flash| hybrid | iterative | cascaded | best | Speed Up |
|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 1056.8 | 393.2 | 14.0 | NA | 22.6 | 23.9 | 24.2 | 21.0 | Select | 1.0x |
| 2 | 1057.7 | 435.0 | 31.1 | NA | 22.7 | 24.6 | 24.3 | 20.9 | Cascaded | 1.5x |
| 4 | 1058.2 | 358.7 | 52.6 | NA | 22.8 | 24.0 | 24.3 | 20.9 | Cascaded | 2.5x |
| 5 | 1059.1 | 374.1 | 63.5 | NA | 23.4 | 24.6 | 24.2 | 21.6 | Cascaded | 2.9x |
| 8 | 1057.6 | 392.4 | 96.1 | NA | 23.3 | 24.5 | 24.6 | 21.6 | Cascaded | 4.5x |
| 16 | 1056.9 | 384.3 | 184.0 | NA | 24.1 | 29.4 | 26.8 | 23.0 | Cascaded | 8.0x |
| 32 | 1060.1 | 363.9 | 357.1 | NA | 26.0 | 28.0 | 29.0 | 28.2 | Flash | 13.7x |
| 50 | 1057.3 | 440.4 | 552.3 | NA | 28.1 | 29.7 | 31.4 | 31.0 | Flash | 19.7x |
| 64 | 1060.6 | 378.2 | 706.4 | NA | 29.4 | 29.7 | 31.4 | 31.3 | Flash | 24.0x |
| 100 | 1057.2 | 364.5 | 1097.0 | NA | NA | 32.3 | NA | NA | Hybrid | 32.7x |
| 128 | 1057.2 | 362.2 | 1401.3 | NA | NA | 32.7 | NA | NA | Hybrid | 32.4x |
| 256 | 1059.1 | 346.8 | 2788.2 | NA | NA | 36.4 | NA | NA | Hybrid | 29.1x |

As shown, while distributed select sort is optimal for k up to 4 at batch size 1, the new specialized kernels provide a dramatic, order-of-magnitude performance improvement for larger k values common in LLM workloads.

Note that unstable sort is used by default for fast top k kernels. You can define STABLE_TOPK during compilation to use stable sort for testing purpose.

### CODE REVIEW

File: src/cuda/cuda_topk.h [hunk 1]
```diff
@@ -39,10 +67,15 @@ struct TopkData {
   // - Full sort: Stores the start offset of each batch segment for CUB's segmented sort
   cuda_unique_ptr<int> batch_offsets;
 
+  // --- Buffers specific to Distributed Sort ---
+  cuda_unique_ptr<int> top_k_distributed_lock;
```
Comment (hariharans29) [inline src/cuda/cuda_topk.h]: Remove Distributed Sort relevant code as it will come in a separate PR ?

File: src/cuda/cuda_topk_benchmark.cuh [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      });
+      if (flash_latency < min_latency) {
+        min_latency = flash_latency;
+        best_algo = TopkAlgo::FLASH;
+      }
+    }
+  }
+
+  // Cache the result for future calls to avoid re-benchmarking.
+  topk_data->best_algo_cache[k] = best_algo;
```
Comment (tianleiwu) [inline src/cuda/cuda_topk_benchmark.cuh]: TODO: cache result globally so that we need not run it twice for one device.

File: src/cuda/cuda_sampling.cu [hunk 3]
```diff
@@ -10,6 +10,7 @@
 #include <iostream>
 #include <limits>
 #include <stdio.h>
+#include "../generators.h"
```
Comment (kunal-vaishnavi) [inline src/cuda/cuda_sampling.cu]: Why does this header need to be included?

File: src/cuda/cuda_topk.cu [hunk 4]
```diff
@@ -3,30 +3,106 @@
 
 #include <cub/device/device_segmented_radix_sort.cuh>
 
+#include "../generators.h"
```
Comment (kunal-vaishnavi) [inline src/cuda/cuda_topk.cu]: Same question

File: src/cuda/cuda_topk.h [hunk 5]
```diff
@@ -3,68 +3,172 @@
...
+#endif
+
+// Helper to align memory addresses.
+__host__ __device__ inline size_t AlignUp(size_t size, size_t alignment) {
+  return (size + alignment - 1) & ~(alignment - 1);
+}
+
+constexpr int kGpuBufferAlignment = 256;
+constexpr int kHybridSortMaxK = 256;         // The maximum k (up to 256) allowed for hybrid sort. Must be power of 2.
+constexpr int kFlashSortMaxK = 128;          // The maximum k (up to 256) allowed for flash sort. Must be power of 2.
```
Comment (kunal-vaishnavi) [inline src/cuda/cuda_topk.h]: Can you add comments on how these values have been determined?

**Comment by hariharans29** [comment]
Is https://github.com/microsoft/onnxruntime-genai/pull/1710 relevant for any combination of {bs, vocab_size, k} after this change ?

File: src/cuda/cuda_topk.h [hunk 6]
```diff
@@ -3,68 +3,172 @@
...
+constexpr int kLlmSortMaxK = 64;             // The maximum k (up to 256) allowed for LLM sort. Must be power of 2.
+constexpr int kMaxBenchmarkLocalCache = 64;  // The maximum local cache of online benchmarking results.
+
+// Enum for the different Top-K algorithms used in online benchmarking.
+enum class TopkAlgo { SELECTION,
```
Comment (hariharans29) [inline src/cuda/cuda_topk.h:55]: Maybe provide brief intuition regarding which algo performs best for which problem space (like the one given in the PR description) ?

File: src/cuda/cuda_sampling.cu (hunk 3 above)
Comment (tianleiwu) [reply src/cuda/cuda_sampling.cu]: no. removed.

File: src/cuda/cuda_topk.cu (hunk 4 above)
Comment (tianleiwu) [reply src/cuda/cuda_topk.cu]: removed.

**Comment by hariharans29** [comment]
If some algos are using unstable sort by default- isn't that a breaking change ?

**Comment by tianleiwu** [comment]
> If some algos are using unstable sort by default- isn't that a breaking change ?

It's not a breaking change for LLM usage. Sampling for LLM is not deterministic: you will get different results if seed is different.

If there some candidates with same logits (same probabilities), it is better let each candidate have same chance to be sampled. If we use stable sort (tie-breaking), that mean we give the token with smaller token id to have higher chance to be selected in top k especially when they compete for the k-th position. 

Unstable sort means that when token x and y at positions k or k+1, we let sorting algorithm to choose which one wins. So this is a part of sampling.

Current code have choice to build binary: either stable sort or unstable sort. If uses them to do end-to-end evaluation. I guess results shall be close enough.

**Review by kunal-vaishnavi — DISMISSED** [review]

File: src/cuda/cuda_topk_distributed_select_sort.cuh [hunk 7]
```diff
@@ -169,25 +171,32 @@ void LaunchGetDistributedSelectSortTopK(cudaStream_t stream, float* scores_in, f
                                                                              distributed_scores_out);
 }
 
-void RunTopKViaDistributedSelectionSort(TopkData* data, cudaStream_t stream, const float* scores_in, int vocab_size, int k) {
-  float* topk_scores = data->intermediate_scores_1.get();
-  int* topk_indices = data->intermediate_indices_1.get();
+void RunTopK(TopkData* data, cudaStream_t stream, const float* scores_in, int vocab_size, int batch_size, int k) {
```
Comment (hariharans29) [inline src/cuda/cuda_topk_distributed_select_sort.cuh]: Could you please make the float -> int change in your PR ?

https://github.com/microsoft/onnxruntime-genai/pull/1710#discussion_r2360580307

**Review by hariharans29 — APPROVED** [review]
